resource-managers/kubernetes/token-refresh-server/src/main/scala/org/apache/spark/security/kubernetes/TokenRefreshService.scala

```diff
@@ -0,0 +1,349 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.security.kubernetes
+
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream, DataInputStream, DataOutputStream}
+import java.security.PrivilegedExceptionAction
+import java.util.concurrent.TimeUnit
+import java.util.regex.Pattern
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.concurrent.duration.Duration
+import scala.concurrent.ExecutionContext.Implicits.global
+import akka.actor.{Actor, ActorRef, ActorSystem, Cancellable, Props}
+import io.fabric8.kubernetes.api.model.{ObjectMeta, Secret}
+import io.fabric8.kubernetes.client.KubernetesClient
+import org.apache.commons.codec.binary.Base64
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.FileSystem
+import org.apache.hadoop.hdfs.security.token.delegation.DelegationTokenIdentifier
+import org.apache.hadoop.security.token.delegation.AbstractDelegationTokenIdentifier
+import org.apache.hadoop.security.{Credentials, UserGroupInformation}
+import org.apache.hadoop.security.token.{Token, TokenIdentifier}
+import org.apache.spark.security.kubernetes.constants._
+
+
+private class TokenRefreshService(kubernetesClient: KubernetesClient) extends Actor with Logging {
+
+  private val scheduler = context.system.scheduler
+  private val secretUidToTaskHandle = mutable.HashMap[String, Cancellable]()
```

Comment:
No, it will be accessed only by a single thread in charge of this actor. None of these fields are accessed by multiple threads, which is the advantage of using the actor model.